kernel:  Replace redundant switch_handle assignment with assertion

TaiJuWu opened the pull request:
The switch_handle for the outgoing thread is expected to be NULL or dummy thread at the start of a context switch. The previous code performed a redundant assignment to NULL.

This change replaces the assignment with an __ASSERT(). This makes the code more robust by explicitly enforcing this precondition, helping to catch potential scheduler bugs earlier.

andyross reviewed (commented):
Why is the dummy check needed?  That sounds like an uninitialized field we should just fix (for historical reasons we don't actually bzero the dummy thread struct and have a custom "initialize a dummy thread" function that leaves holes).

Generally agree that this looks great as it eliminates a store on a performance path.

TaiJuWu committed:
kernel:  Replace redundant switch_handle assignment with assertion

The switch_handle for the outgoing thread is expected to be NULL
at the start of a context switch.
The previous code performed a redundant assignment to NULL.

This change replaces the assignment with an __ASSERT(). This makes the
code more robust by explicitly enforcing this precondition, helping to
catch potential scheduler bugs earlier.

Also, the switch_handle pointer is used to check a thread's state during a
context switch. For dummy threads, this pointer was left uninitialized,
potentially holding a unexpected value.

Set the handle to NULL during initialization to ensure these threads are
handled safely and predictably.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

TaiJuWu commented on kernel/sched.c:
```diff
@@ -856,7 +856,11 @@ void *z_get_next_switch_handle(void *interrupted)
+
+		__ASSERT(old_thread->switch_handle == NULL ||
+		is_thread_dummy(old_thread->switch_handle),
+		"old thread handle should be null or dummy thread.");
```
Thanks for your point out, fixed these two issues.
